opensquirrel/circuit.py

```diff
@@ -84,27 +84,25 @@ def number_of_qubits(self) -> int:
     def qubit_register_name(self) -> str:
         return self.squirrel_ir.qubit_register_name
 
-    def decompose_mckay(self):
-        """Perform gate fusion on all one-qubit gates and decompose them in the McKay style.
-
-        * all one-qubit gates on same qubit are merged together, without attempting to commute any gate
-        * two-or-more-qubit gates are left as-is
-        * merged one-qubit gates are decomposed according to McKay decomposition, that is:
-                gate   ---->    Rz.Rx(pi/2).Rz.Rx(pi/2).Rz
-        * _global phase is deemed irrelevant_, therefore a simulator backend might produce different output
-            for the input and output circuit - those outputs should be equivalent modulo global phase.
+    def merge_single_qubit_gates(self):
+        """Perform gate fusion on all one-qubit gates.
+        Gates obtained from merging other gates become anonymous gates.
         """
 
-        self.squirrel_ir = mckay_decomposer.decompose_mckay(self.squirrel_ir)  # FIXME: inplace
+        merger.merge_single_qubit_gates(self.squirrel_ir)
 
-    def replace(self, gate_name: str, f):
-        """Manually replace occurrences of a given gate with a list of gates.
+    def decompose(self, decomposer: Callable[[Gate], List[Gate]]):
+        """Generic decomposition pass. It applies the given decomposer function
+        to every gate in the circuit."""
+        replacer.decompose(self.squirrel_ir, decomposer)
```

Comment:
Done! I've created a `Decomposer` ABC with a `decompose` `@abstractmethod`

I could make it a `@staticmethod` as well. This is the most common use case. McKay decomposer doesn't need state
However, we could imagine a decomposer with a state? I'm not sure if this abstract decompose method should be static or not...

In the current change it is not static, so there is this `self` keyword that we don't really need for now.

Reply:
Yes, the way it is used, it is indeed a static method. It would only not be static if we would want to create several instances of a, e.g. McKay, decomposer and call decompose on these objects; but I don't think this is what we want to do.

Reply:
So I'll make the abstract method @static for now and we can reconsider later should we need it to not be static

Reply:
I think it's clearer as static otherwise it might confuse people as to what the role of a decomposer is..